### PR TITLE
NeMo Distillation API

### PR DESCRIPTION
**Pull request by AAnoosheh**
# What does this PR do ?

Add Logits-only Distillation ability to NeMo GPT models.

**Collection**: NLP

# Changelog 
- Full-fledged example script containing a distillation-enabled subclass of `MegatronGPTModel` and proper loss function for logit-based knowledge distillation.
- Default config file, docs, and CICD test

# Usage

```bash
STUDENT_CKPT=".../model1.nemo"  # can also be None
TEACHER_CKPT=".../model2.nemo"
TOKENIZER=".../tokenizer.model"
FINAL_SAVE_FILE="final_checkpoint.nemo"
TP=4
PP=1
NPROC=$((TP * PP))
launch_config="torchrun --nproc_per_node=$NPROC"

${launch_config} examples/nlp/language_modeling/megatron_gpt_distillation.py \
    model.restore_from_path=$STUDENT_CKPT \
    model.kd_teacher_restore_from_path=$TEACHER_CKPT \
    model.tensor_model_parallel_size=${TP} \
    model.pipeline_model_parallel_size=${PP} \
    model.micro_batch_size=1 \
    model.tokenizer.model=$TOKENIZER \
    model.nemo_path=$FINAL_SAVE_FILE \
    trainer.precision=bf16 \
    trainer.devices=$NPROC
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [x] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 1]
```diff
@@ -1000,6 +1013,11 @@ def save_to(self, model, save_path: str):
 
         dist_ckpt_dir = None
 
+        # [ModelOpt]: If Distillation, export the student model.
+        if HAVE_MODELOPT and isinstance(unwrap_model(model), mtd.DistillationModel):
```
Comment (ChenhanYu) [inline nemo/collections/nlp/parts/nlp_overrides.py:1018]: `unwrap_model(model, (Float16Module, MCoreFloat16Module))`

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 2]
```diff
@@ -1325,8 +1343,20 @@ def dummy():
                         instance.get_model_module_list(), tmp_model_weights_dir, prefix="model."
                     )
 
+                # [ModelOpt]: Initial loading from non-resume sharded checkpoint to a Distillation Model
+                # will result in key mismatch with loss modules potentially containing parameters, since
+                # it requires generating a state_dict before loading. Here we hide those modules if present.
+                # Context managers not activated if `instance.model` inner class not a DistillationModel.
+                with ExitStack() as stack:  # Allows multiple context managers for each model shard
+                    model = instance.model if isinstance(instance.model, list) else [instance.model]
+                    model = unwrap_model(model)
```
Comment (ChenhanYu) [inline nemo/collections/nlp/parts/nlp_overrides.py:1353]: `unwrap_model(model, (Float16Module, MCoreFloat16Module))`

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 3]
```diff
@@ -533,7 +535,18 @@ def load_checkpoint(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
...
                 )
 
-            sharded_state_dict = self.lightning_module.sharded_state_dict()
+            # [ModelOpt]: Initial loading from non-resume sharded checkpoint to a Distillation Model
+            # will result in key mismatch with loss modules potentially containing parameters, since
+            # it requires generating a state_dict before loading. Here we hide those modules if present.
+            # Context managers not activated if `self.model` inner class not a DistillationModel.
+            with ExitStack() as stack:  # Allows multiple context managers for each model shard
+                model = self.model if isinstance(self.model, list) else [self.model]
+                model = unwrap_model(model)
```
Comment (ChenhanYu) [inline nemo/collections/nlp/parts/nlp_overrides.py:545]: `unwrap_model(model, (Float16Module, MCoreFloat16Module))`

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 4]
```diff
@@ -533,7 +535,18 @@ def load_checkpoint(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
                     self.lightning_module.get_model_module_list(), checkpoint_path, prefix="model."
                 )
 
-            sharded_state_dict = self.lightning_module.sharded_state_dict()
+            # [ModelOpt]: Initial loading from non-resume sharded checkpoint to a Distillation Model
```
Comment (ChenhanYu) [inline nemo/collections/nlp/parts/nlp_overrides.py:539]: Since we only enable logit-distillation now, we don't have extra parameters right?. If so, let's remove this special handling here.

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 5]
```diff
@@ -1325,8 +1343,20 @@ def dummy():
                         instance.get_model_module_list(), tmp_model_weights_dir, prefix="model."
                     )
 
+                # [ModelOpt]: Initial loading from non-resume sharded checkpoint to a Distillation Model
```
Comment (ChenhanYu) [inline nemo/collections/nlp/parts/nlp_overrides.py:1347]: Same above. We don't need this for logit-only-distillation.

**Review by ChenhanYu — COMMENTED** [review]
Left some comments

File: examples/nlp/language_modeling/megatron_gpt_distillation.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+from nemo.utils.model_utils import load_config, unwrap_model
+
+
+mp.set_start_method("spawn", force=True)
+
+
+class DistillationMegatronGPTModel(MegatronGPTModel):
+    """..."""
+
+    def __init__(self, *args, **kwargs):
```
Comment (ChenhanYu) [inline examples/nlp/language_modeling/megatron_gpt_distillation.py:45]: This has to be (cfg, trainer); otherwise error out.

File: examples/nlp/language_modeling/megatron_gpt_distillation.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+        """Constructor."""
+        super().__init__(*args, **kwargs)
```
Comment (ChenhanYu) [inline examples/nlp/language_modeling/megatron_gpt_distillation.py:47]: This has to be (cfg, trainer); otherwise error out.

File: examples/nlp/language_modeling/megatron_gpt_distillation.py
Comment (ericharper) [inline examples/nlp/language_modeling/megatron_gpt_distillation.py:1]: This file needs copyright header

Comment (AAnoosheh) [reply examples/nlp/language_modeling/megatron_gpt_distillation.py:1]: Done. thanks

File: docs/source/nlp/distillation.rst [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+Distillation
+==========================
+
+Knowledge Distillation (KD)
+--------------------------------
+
+KD involves using information from an existing trained model to train a second (usually smaller, faster) model, thereby "distilling" knowledge from one to the other.
+
+Distillation has two primary benefits: potentially faster convergence and/or higher end accuracy than traditional training.
```
Comment (sharathts) [inline docs/source/nlp/distillation.rst:11]: I think we can just say "faster convergence and higher end accuracy than traditional training"

File: docs/source/nlp/distillation.rst [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+In NeMo, distillation is enabled by the `NVIDIA TensorRT Model Optimizer (ModelOpt) <https://github.com/NVIDIA/TensorRT-Model-Optimizer>`_ library -- a library to optimize deep-learning models for inference on GPUs.
+
+The logits-distillation process consists of the following steps:
+
+1. Loading both student and teacher model checkpoints using an appropriate parallelism strategy
```
Comment (sharathts) [inline docs/source/nlp/distillation.rst:17]: perhaps mention student and teacher need to have identical parallelism strategies?

File: docs/source/nlp/distillation.rst [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+2. Training until convergence, where forward passes are run on both models, performing a specific loss function between the logits.
```
Comment (sharathts) [inline docs/source/nlp/distillation.rst:18]: fwd only on teacher, fwd+bwd+update on student

File: examples/nlp/language_modeling/megatron_gpt_distillation.py [hunk 11]
```diff
@@ -0,0 +1,577 @@
...
+def load_distillation_config(student_cfg: TransformerConfig) -> Dict[str, Any]:
+    """Create a default distillation config for MCore GPT Models.
+
+    Args:
+        student_cfg: Model config for student model.
+    """
+    logit_pair = ("output_layer", "output_layer")  # logit module names for MCoreGPTModel
+    tp_enabled = student_cfg.tensor_model_parallel_size > 1
+
+    cfg = {
```
Comment (sharathts) [inline examples/nlp/language_modeling/megatron_gpt_distillation.py:311]: Can these settings also moved to `config.yaml`?

Comment (AAnoosheh) [reply examples/nlp/language_modeling/megatron_gpt_distillation.py:311]: I would rather leave it here for now since the name of the logits layer is fixed for GPT models, and the TP thing is just a check.

Then user has fewer things to accidentally not touch.

Comment (sharathts) [reply examples/nlp/language_modeling/megatron_gpt_distillation.py:311]: Discussed to move it to config once intermediate distillation support is added.

**Review by ChenhanYu — APPROVED** [review]
LGTM; Let's see what's remaining opens

**Comment by AAnoosheh** [comment]
@pablo-garay @ko3n1g  I'd like to request a review, thanks!

**Review by ko3n1g — APPROVED** [review]

**Comment by AAnoosheh** [comment]
@pablo-garay is OOTO and don't know til when.

@ko3n1g You're a code owner for same file, so it's redundant.  Could we bypass Pablo so we can merge?

**Comment by ko3n1g** [comment]
hey @AAnoosheh, I'm still working on a fix for the CI. Unless it's highly business critical I would prefer waiting until our infra is up again such that we fully test this PR

**Comment by AAnoosheh** [comment]
@ChenhanYu How business critical is this atm?

**Review by sharathts — APPROVED** [review]
LGTM

**Review by ko3n1g — APPROVED** [review]

**Comment by sangHa0411** [comment]
Hi I want to use this code for Distillation.
I was not able to find **hide_teacher_model** function in "examples/nlp/language_modeling/megatron_gpt_distillation.py" and I got error.

Where can I find **hide_teacher_model** function?

**Comment by AAnoosheh** [comment]
Hi, @sangHa0411 

The `hide_teacher_model` is part of the Distillation module in the [Model Optimizer library](https://github.com/NVIDIA/TensorRT-Model-Optimizer) but not officially publicized yet. 

If you have v0.15 of the library installed and use the script as is, it should be present in the model after it gets converted – on [this line](https://github.com/NVIDIA/NeMo/blob/d7d6b99c22d3edaeb0f61b2b1719e4e400f540af/examples/nlp/language_modeling/megatron_gpt_distillation.py#L160)

**Comment by lyravv** [comment]
“Pipeline parallelism not yet supported”，then how to deal with the models like llama 70B? Do we apply the tensor parallelism 
 across multiple nodes?

**Comment by AAnoosheh** [comment]
@lyravv
Llama-70b can't even be tensor-parallel split further than 8 devices as it has GQA with 8 groups.
It would require shard duplication, and I'm not sure if NeMo already supports that or not.

But PP support is underway and hopefully available soon.

**Comment by zhaoyang-star** [comment]
Now the TP has much less TFLPs than PP. So PP support is very important and I am very interest in it.

**Comment by zhaoyang-star** [comment]
By the way, how to print the `ce_loss` and `kd_loss` before combine them together? I tried to add `print` in /usr/local/lib/python3.10/dist-packages/modelopt/torch/distill/loss_balancers.py at `StaticLossBalancer::forward`, but it seems the logging is disabled and no output. Does anyone has similar issue? Thanks.
